src/librustdoc/passes/doc_test_lints.rs

```diff
@@ -67,13 +67,28 @@ crate fn should_have_doc_example(cx: &DocContext<'_>, item: &clean::Item) -> boo
                 | clean::ImportItem(_)
                 | clean::PrimitiveItem(_)
                 | clean::KeywordItem(_)
+                // check for trait impl
+                | clean::ImplItem(clean::Impl { trait_: Some(_), .. })
         )
     {
         return false;
     }
+
     // The `expect_def_id()` should be okay because `local_def_id_to_hir_id`
     // would presumably panic if a fake `DefIndex` were passed.
     let hir_id = cx.tcx.hir().local_def_id_to_hir_id(item.def_id.expect_def_id().expect_local());
+
+    // check if parent is trait impl
+    if let Some((_, parent_node)) = cx.tcx.hir().parent_iter(hir_id).next() {
```

Comment:
I think only checking for the parent of a node is enough, but I am unsure if there is any edge case where an item more than immediately below a trait impl could run through here.

Reply:
I don't think so, unless you're documenting inside an item in a trait impl.

Reply:
[get_parent_node](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.get_parent_node) might be better in this case:

```suggestion
    if let Some((_, parent_node)) = cx.tcx.hir().get_parent_node(hir_id) {
```

Reply:
So [`get_parent_node` just calls `find_parent_node` and returns the crate id if nothing is found](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#308-310) and returns a `HirId` and thus makes another `cx.tcx.hir().find` call necessary, which is basically what [`ParentHirIterator::next` does](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.ParentHirIterator.html#method.next).

Overall this might be more efficient but harder to read:
```rust
    if let Some(parent_node) =
        cx.tcx.hir().find_parent_node(hir_id).and_then(|id| cx.tcx.hir().find(id))
    {
        if matches!(
            parent_node,
            hir::Node::Item(hir::Item {
                kind: hir::ItemKind::Impl(hir::Impl { of_trait: Some(_), .. }),
                ..
            })
        ) {
            return false;
        }
    }
```

I am guessing the more efficient version is generally preferred?

Reply:
I actually prefer the `find_parent_node` version, I find it more explicit. And if it's the most performant, it's even better! :)